Add `SizeContainer`

StarryWorm opened the pull request:
* Closes https://github.com/godotengine/godot-proposals/issues/8333
* Supersedes/Implements internally https://github.com/godotengine/godot-proposals/issues/13534
* Supersedes #94171
* Includes the fix in #112525 (Noted such that if that PR causes any issues, the fix to said issues gets ported here)
  * Note: This will not be true in the final version of the PR, as #112741 has replaced #112525 as a much more elegant solution, and this PR will thus adapt around that. 

Adds a `custom_maximum_size` internal property to `Control`. The `maximum_size` will serve as a cap on the container's size. 

The `custom_maximum_size` is a `Size2`, and will apply dimension-wise if that dimension's value is greater than 0.
* e.g. `custom_maximum_size= (100,0)` will cap the width to 100 but will not cap the height

~ 

Adds a `SizeContainer` which exposes the `custom_maximum_size` property. This container will also apply this property to all of its children Controls. 
If a child has the `Expand` size flag, the container will expand to its maximum size in the specified axis.
Cannot have specific anchor flags, which would cause it to expand to be larger than its designated maximum size (namely, all of the Wide types and the Full Rect). This can technically still be overridden by using Custom anchoring, but that's A. beyond my control, B. an advanced-user thing -- use it at your own risk.

Additional feature:
* A `ScrollContainer` with a `custom_maximum_size` (i.e. within a `SizeContainer`) will first expand as much as it can until it reaches the maximum size, and then will start showing scrollbars as needed. 

Future work:
* Expose the horizontal component of `custom_maximum_size` in `Label` as `autowrap_width`, implementing https://github.com/godotengine/godot-proposals/issues/13568

AThousandShips commented on doc/classes/SizeContainer.xml:
```diff
@@ -0,0 +1,31 @@
+	</brief_description>
+	<description>
+		[SizeContainer] allows setting a maximum size for itself. When the container's size exceeds the defined maximum size, it will not grow any larger, effectively capping its dimensions. This is useful for creating UI elements that should not exceed a certain size regardless of their content or parent container's size.
+		To control the [SizeContainer]'s maximum size, use the [code skip-lint]max_size[/code] property as shown below.
```
```suggestion
		To control the [SizeContainer]'s maximum size, use the [member maximum_size] property as shown below.
```

KoBeWi commented:
I'm using the latest version and I do have maximum size set. Can you share your scene?

KoBeWi commented:
I found the difference. My ScrollContainer had `vertical_scroll_mode` set to Auto. I noticed your ScrollContainer has `horizontal_scroll_mode` disabled, so I did that (for vertical) and it works. It's weird though, why is orthogonal scroll mode relevant, if the container grows in a different direction?

StarryWorm commented:
Well, that's definitely unintended behavior. I'll take a look at why it's happening.

Edit: I had it disabled because getting the horizontal scrollbar was annoying me, which thinking about it, it shouldn't have showed up in the first place :/

StarryWorm committed:
Add `SizeContainer`

Adds a `SizeContainer` node, which supports modifying the Control `custom_maximum_size` property.
If a child has expand size flag, the container will expand to its maximum size in the specified axis
Cannot have specific anchor flags which would cause it to expand to be larger than its designated maximum size

StarryWorm committed:
Add `custom_maximum_size` property to `Control`

The `maximum_size` will serve as a cap on the size of the container. All children of this container will be clipped to fit within that cap.

Adapt `ScrollContainer` to work with new `custom_maximum_size`

AThousandShips commented on scene/gui/control.cpp:
```diff
@@ -1649,32 +1649,43 @@ Vector2 Control::get_combined_pivot_offset() const {
+
+	if (minsize != data.last_minimum_size || maxsize != data.last_maximum_size) {
+		Size2 old_minsize = data.last_minimum_size;
+		if ((data.size_cache.is_equal_approx(old_minsize)) && (minsize.width < old_minsize.width || minsize.height < old_minsize.height)) {
```
```suggestion
		if (data.size_cache.is_equal_approx(old_minsize) && (minsize.width < old_minsize.width || minsize.height < old_minsize.height)) {
```
Confusing parentheses

AThousandShips commented on doc/classes/SizeContainer.xml:
```diff
@@ -0,0 +1,28 @@
+		[csharp]
+		// This code sample assumes the current script is extending SizeContainer.
+		Vector2 maxSizeValue = new Vector2(400, 300);
+		SetCustomMaximumSize(maxSizeValue);
```
This is not valid C#, setters are not exposed

AThousandShips commented on scene/gui/size_container.cpp:
```diff
@@ -0,0 +1,169 @@
+	}
+
+	control->set_custom_maximum_size(Size2());
+	control->disconnect("size_flags_changed", Callable(this, "_child_size_bounds_changed"));
```
```suggestion
	control->disconnect(SceneStringName(size_flags_changed), Callable(this, "_child_size_bounds_changed"));
```

AThousandShips commented on doc/classes/Control.xml:
```diff
@@ -1018,6 +1045,11 @@
+		<member name="custom_maximum_size" type="Vector2" setter="set_custom_maximum_size" getter="get_custom_maximum_size" default="Vector2(0, 0)">
+			The maximum size of the node's bounding rectangle. If set to a value greater than [code](0, 0)[/code], the node's bounding rectangle will never exceed this size. A value of [code](0, 0)[/code] means there is no maximum size.
+					[b]Note:[/b] This property is only accessible in code and not in the Inspector.
+					[b]Note:[/b] Not all [Control] subtypes handle this gracefully, which may lead to unexpected behavior if the control's contents exceed this size.
```
Incorrect formatting for these two lines

AThousandShips commented on doc/classes/Control.xml:
```diff
@@ -963,7 +984,13 @@
+		<method name="update_size_bounds">
 			<return type="void" />
 			<description>
 				Invalidates the size cache in this node and in parent nodes up to top level. Intended to be used with [method get_minimum_size] when the return value is changed. Setting [member custom_minimum_size] directly calls this method automatically.
```
Should mention the new maximum size

KoBeWi commented:
Seems like most changes here come from method rename?
The combination of both approaches would be doing the same as ScrollContainer's `max_size`, but making the property only available via SizeContainer. Which is awkward tbh.

ohboyotero commented:
> Personally, I actually believe the former (namely `Container`-level handling) is better.

Agreed. I understand the "one container type, one function" ideology, but there are a few things that just make so much sense to push to the root imho as they are so often required together.  This (min_size, max_size, "overflow mode") feels like one of them. Personally, I also find myself adding inner/outer `MarginContainer`s _everywhere_.

Similar sentiment from @shadowblitz16 here: https://github.com/godotengine/godot-proposals/issues/8333#issuecomment-3523359431

> we're just creating a fancy ScrollContainer

Yeah, that makes sense. If the max_size functionality does end up having to be merged into ScrollContainer, it would be great to rename ScrollContainer accordingly to reflect the greater functionality (e.g., `OverflowContainer` or `ConstrainedContainer` or something), but I imagine that is far too big/breaking a change.

I hate this suggestion tbh, but maybe just add a type alias with the new name? That way the larger functionality is at least discoverable in scanning through the docs/types, and the core intent behind use of the component can be clarified by selecting the more appropriate name for code readability?

StarryWorm commented:
Closing this PR in favor of #116640, which is a much cleaner implementation and has a lot less baggage than this one. 
Will reopen if, for some reason, this solution is favored (the general sentiment from the contributors chat seems to be that the new implementation is the correct path).
